### ISSUE TITLE
Add ASYNC_CACHE for alluxio.user.file.writetype.default

### ISSUE THREAD

**Issue by JySongWithZhangCe** [Alluxio/alluxio#15550]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when I want the task to be done quickly after the data is written to UFS, and this part of the data will be automatically cached in Alluxio.

**Describe the solution you'd like**
Add ASYNC_CACHE write mode that write to UFS first, and async load data into Alluxio.

**Describe alternatives you've considered**
Nope.

**Urgency**
Urgent.

**Additional context**
There are more benefits with ASYNC_CACHE. Users can choose which workers to cache data to avoid performance degradation of distributed computing tasks due to load imbalance of individual workers.

**Comment by jiacheliu3** [comment]
If you are comparing your ASYNC_CACHE to CACHE_THROUGH, note that Alluxio builds on the assumption that cache write is faster than UFS write, so whenever UFS write finishes, the cache write has already finished. I'd guess in most cases making the cache write async would not give you a better performance. Another important thing is, Alluxio assumes the alluxio-UFS bandwidth is expensive, while the client-alluxio bandwidth is much cheaper as we deploy Alluxio and the app co-located. Loading from UFS is more expensive than just write to cache.

**Comment by JySongWithZhangCe** [comment]
> If you are comparing your ASYNC_CACHE to CACHE_THROUGH, note that Alluxio builds on the assumption that cache write is faster than UFS write, so whenever UFS write finishes, the cache write has already finished. I'd guess in most cases making the cache write async would not give you a better performance. Another important thing is, Alluxio assumes the alluxio-UFS bandwidth is expensive, while the client-alluxio bandwidth is much cheaper as we deploy Alluxio and the app co-located. Loading from UFS is more expensive than just write to cache.

I agree that alluxio is designed to better accelerate remote UFS, and at the same time, high-reliability and efficient computing based on local UFS will also be a very common method. Coordinating local UFS and remote UFS to build a high-performance, reliable, and low-cost data analysis system is one of Alluxio's strengths.

**Comment by JySongWithZhangCe** [comment]
> If you are comparing your ASYNC_CACHE to CACHE_THROUGH, note that Alluxio builds on the assumption that cache write is faster than UFS write, so whenever UFS write finishes, the cache write has already finished. I'd guess in most cases making the cache write async would not give you a better performance. Another important thing is, Alluxio assumes the alluxio-UFS bandwidth is expensive, while the client-alluxio bandwidth is much cheaper as we deploy Alluxio and the app co-located. Loading from UFS is more expensive than just write to cache.

The data write code shows that CACHE_THROUGH will write the data twice, which significantly increases the time of the data write. https://github.com/Alluxio/alluxio/blob/master/core/client/fs/src/main/java/alluxio/client/file/AlluxioFileOutStream.java
```
  private void writeInternal(int b) throws IOException {
    if (mShouldCacheCurrentBlock) {
      try {
        if (mCurrentBlockOutStream == null || mCurrentBlockOutStream.remaining() == 0) {
          getNextBlock();
        }
        mCurrentBlockOutStream.write(b);
      } catch (IOException e) {
        handleCacheWriteException(e);
      }
    }

    if (mUnderStorageType.isSyncPersist()) {
      mUnderStorageOutputStream.write(b);
      Metrics.BYTES_WRITTEN_UFS.inc();
    }
    mBytesWritten++;
  }
```
After verification in production scenarios, the time-consuming of writing data only to the Alluxio cache (MUST_CACHE) is roughly the same as that of writing directly to HDFS, and the secondary writing of data will significantly affect job performance. And when the cache of individual nodes is full, data eviction further slows down the data writing performance. According to user feedback, the operating efficiency of their jobs in this case is reduced by 50% compared to when the cache is sufficient.

**Comment by jiacheliu3** [comment]
@JySongWithZhangCe Good point. The client writing twice (to UFS stream and to worker) is a problem that Alluxio needs to solve. The two streams should be able to merge into one, then we will not have the double-write problem here.

Are you saying in your environment MUST_CACHE speed is the same to THROUGH? If that's true then it suggests either:
(1) Your alluxio cache is not fast enough, you are probably using HDD to serve cache. In order to have cache speedup you need to use SSD for cache. 
(2) Your HDFS is probably colocated with the compute application, or the network speed to HDFS is fast. 

If either of the above is true, I actually don't think ASYNC_CACHE will help you because you may not need cache at all? You can just write with THROUGH.

Regarding cache eviction being slow, which version of Alluxio are you using?

FYI @beinan  @HelloHorizon WDYT?

**Comment by beinan** [comment]
> @JySongWithZhangCe Good point. The client writing twice (to UFS stream and to worker) is a problem that Alluxio needs to solve. The two streams should be able to merge into one, then we will not have the double-write problem here.
> 
> Are you saying in your environment MUST_CACHE speed is the same to THROUGH? If that's true then it suggests either: (1) Your alluxio cache is not fast enough, you are probably using HDD to serve cache. In order to have cache speedup you need to use SSD for cache. (2) Your HDFS is probably colocated with the compute application, or the network speed to HDFS is fast.
> 
> If either of the above is true, I actually don't think ASYNC_CACHE will help you because you may not need cache at all? You can just write with THROUGH.
> 
> Regarding cache eviction being slow, which version of Alluxio are you using?
> 
> FYI @beinan @HelloHorizon WDYT?

I agree with @jiacheliu3, either 1 or 2, I think using THROUGH might help.

**Comment by JySongWithZhangCe** [comment]
@jiacheliu3 @beinan My scene is in line with the second scene that Jia Cheng said. I think like Jia Cheng said, optimizing ASYNC_CACHE and CACHE_THROUGH is a better choice, do you have any suggestions?
